### ISSUE TITLE
External-DNS chart doesn't work for AWS ACM certificate validations

### ISSUE THREAD

**Issue by wparad** [bitnami/charts#17378]
### Name and Version

Bitnami/external-dns 6.13.1

### What architecture are you using?

amd64

### What steps will reproduce the bug?

Set up the chart as:

```yaml
kind: Service
apiVersion: v1
metadata:
  name: test-service
  namespace: external-dns
  annotations:
    external-dns.alpha.kubernetes.io/hostname: _validation.example.com.
spec:
  type: ExternalName
  externalName: auth.example.com
```

### What is the expected behavior?

It works

### What do you see instead?

```
"Service/external-dns/test-service dry-run failed, reason: Invalid: Service "test-service" is invalid: spec.externalName: Invalid value: "_validation.example.com": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')"
```

**Comment by wparad** [comment]
I've also linked the issue over here https://github.com/kubernetes-sigs/external-dns/issues/3743 in case there is a solution in K8s core instead.

**Comment by wparad** [comment]
@fevisera @carrodher...the bot made me do it.

**Comment by fevisera** [comment]
Hi @wparad,
Thank you for reporting the error. I could not reproduce your error. Can you provide details on how you got to that error? Did you try it with a different hostname?

**Comment by wparad** [comment]
hostnames without the `_` work, but hostnames with the underscore were causing a problem.

**Comment by fevisera** [comment]
Hi @wparad,

I apologize for the delay in getting back to you.

After reviewing the RFC mentioned in the error, it appears that the usage of underscores for hostnames is restricted. While I'm not an expert on this topic, I found a more detailed explanation on https://stackoverflow.com/a/14622263.

If you believe this is still a valid bug, I kindly request you provide additional details and context regarding the issue. This will help me better understand the problem.

Thank you.

**Comment by wparad** [comment]
That SO post is confusing me, to be honest. I don't really get what's the difference between a "hostname" and a "domain label". For me there is just a string that contains an `_`. According to the post sometimes this is allowed and other times this should not be allowed. Further it seems that the validation in chart is based on the assumption that `_` is never allowed. But from what you are saying you have no problem using this chart to generate the AWS ACM validation records that contain an `_` in the CNAME record `name` and record `value`. Is that correct?

**Comment by fevisera** [comment]
Hi @wparad,

Thank you for reaching back.

> But from what you are saying you have no problem using this chart to generate the AWS ACM validation records that contain an _ in the CNAME record name and record value. Is that correct?

I'm afraid I cannot help you test AWS ACM validation records.

In regards to that post, it mentions:

> The hostname is subject to the restrictions of [RFC 952](https://www.rfc-editor.org/rfc/rfc952) and the [slight relaxation of RFC 1123](https://www.rfc-editor.org/rfc/rfc1123#page-13)

The [RFC 952](https://www.rfc-editor.org/rfc/rfc952) does not include `_` (underscore) which does not seem to be a valid character for a hostname. I believe that is why you are getting that error.

Since you are using a `external-dns` specific annotation to define a hostname, `external-dns.alpha.kubernetes.io/hostname`, I recommend seeking more information about their implementation from [official repository](https://github.com/kubernetes-sigs/external-dns).

**Comment by wparad** [comment]
It's okay, we don't need another test, we've validated it doesn't work given the error message. But it sounds like you are saying that validation is not coming from this repo. That no where in this repo you are validating the record name (this isn't a domain name, so the RFC does not apply). 

Before we go assume the issue is in another place, could you point me to where in this repo the implementation for these properties is defined so we can figure out what the source of the problem is as there is no "stack trace" associated with the validation.

**Comment by fevisera** [comment]
You can find more details about that chart in [bitnami/external-dns](https://github.com/bitnami/charts/tree/main/bitnami/external-dns) folder. I'm uncertain about the specific properties you are seeking, could you be more specific?

Have you tried to set `logLevel` to `debug` to get more information?  That property is explained in the [readme](https://github.com/kubernetes-sigs/external-dns/tree/master/charts/external-dns).

Also, I'm still unable to reproduce your issue. Could you provide the commands you are executing?

**Comment by wparad** [comment]
@scott-david-walker maybe you know the answers to these questions ^

**Comment by scott-david-walker** [comment]
The problem lies in how RFC 1123 is implemented within Kubernetes. All resource names and identifiers must comply with RFC 1123 which means there cannot be an _ in the name. So when we use the kind of 'Service' along with the 'externalName' property, Kubernetes will throw an error that it isn't allowed. (Just try to kubectl apply -f the service file to see the error.)

My understanding was that creating the service with the annotation was the correct way to implement CNAME records in external dns as I don't see any overrides for it in the values.yaml. @fevisera  - Is there a better way to do so? If so, could you point me to the relevant docs?

**Comment by fevisera** [comment]
Hi @wparad and @scott-david-walker,

It appears that `spec.externalName` is missing, so I created #18213 to include it. With the changes introduced by that [PR](https://github.com/bitnami/charts/pull/18213), I was able to reproduce your error by overriding the values as follows:
```
service:
  annotations: 
    external-dns.alpha.kubernetes.io/hostname: _validation.example.com.
  type: ExternalName
  externalName: _auth.example.com
```
```
$ helm install external-dns . -n external-dns --create-namespace -f my-values.yaml
Error: INSTALLATION FAILED: Service "external-dns" is invalid: spec.externalName: Invalid value: "_auth.example.com": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

If you remove those `_`, the custom values are valid:
```
$ kubectl -n external-dns describe svc
Name:              external-dns
Namespace:         external-dns
Labels:            app.kubernetes.io/instance=external-dns
                   app.kubernetes.io/managed-by=Helm
                   app.kubernetes.io/name=external-dns
                   helm.sh/chart=external-dns-6.23.0
Annotations:       external-dns.alpha.kubernetes.io/hostname: validation.example.com.
                   meta.helm.sh/release-name: external-dns
                   meta.helm.sh/release-namespace: external-dns
Selector:          app.kubernetes.io/instance=external-dns,app.kubernetes.io/name=external-dns
Type:              ExternalName
IP Families:       <none>
IP:                
IPs:               <none>
External Name:     auth.example.com
Port:              http  7979/TCP
TargetPort:        http/TCP
Endpoints:         
Session Affinity:  None
Events:            <none>
```

Please review the mentioned [PR](https://github.com/bitnami/charts/pull/18213) and let me know if anything is still missing. Once that PR is merged, this issue will be automatically closed.

**Comment by wparad** [comment]
I'm not following how this would fix the problem.

> I was able to reproduce your error by overriding the values as follows...
...
> If you remove those _, the custom values are valid:

Okay, but those underscores can't be removed because they are generated by ACM. So while it does appear there was a bug here after all of missing the `externalName`, that doesn't fix the problem, right?

Scott asked:
> How can the bitnami/external-dns be used to generate the necessary route53 records for AWS ACM.

But the answer provided feels like:
> You can't.

Is that correct?

**Comment by fevisera** [comment]
Hi @wparad,

As pointed out by @scott-david-walker, it is worth noting that the Kubernetes requirement seems unrelated to the chart implementation. So I do not think that is really an issue/bug to be tackled from the helm chart side. However, if I am mistaken or overlooking any details, I would appreciate your clarification.

Regarding to:
 > How can the bitnami/external-dns be used to generate the necessary route53 records for AWS ACM.

Thanks to @scott-david-walker's [comment](https://github.com/bitnami/charts/issues/17378#issuecomment-1665340504) I noticed `values.yaml` was not including `externalName` property, so that is why it could not be overridden from values. In my last [comment](https://github.com/bitnami/charts/issues/17378#issuecomment-1665779841) I included the steps to configure the ExternalName service. I this [example](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/externalname.md#externalname-service) from the official repo. 

Besides that configuration, I came across [AWS documentation](https://docs.aws.amazon.com/acm/latest/userguide/troubleshooting-DNS-validation.html) (in the "Underscores prohibited by DNS provider" section) that suggests removing underscores from ACM-provided values as a potential solution in some cases. Have you considered giving this a try?

**Comment by wparad** [comment]
If you read further in the AWS documentation you will see this:
> However, the CNAME name parameter must always begin with a leading underscore.

That is the problem here.

Just so we are all on the same page, we agree that
> It is impossible to use the bitnami/external-dns helm chart to solve this problem.

Is that correct? And we are are saying this is because of k8s.

> So I do not think that is really an issue/bug to be tackled from the helm chart side. However, if I am mistaken or overlooking any details

Where exactly should this issue/bug be tracked then?

**Comment by fevisera** [comment]
Hi @wparad,

Sorry for the delay. Yes, that is correct, the problem is related to k8s, so there is nothing we can do to fix it.

As you have already done in the past (https://github.com/kubernetes-sigs/external-dns/issues/3743), I recommend reporting this matter on the official platform.

As the problem is not connected to the chart, I will close this issue.

Thanks for contacting us.
